### ISSUE TITLE
Support for Property Pages

### ISSUE THREAD

**Issue by anthonydotmoe** [rodrigocfd/winsafe#140]
I'm working on writing an MMC Snap-in, and I'd like to use winsafe to handle property sheet pages. All MMC wants is for you to create your property sheet pages using the Win32 `CreatePropertySheetPage` function, then pass the returned `HPROPSHEETPAGE` to the supplied COM interface (IPropertySheetCallback). It seems like property sheet pages are simply modeless dialog boxes presented in a tabular format, so I think that adapting should be simple enough.
I briefly tried extending the library, creating a `PropsheetPage` struct similar to the `RawModeless` struct, but I still haven't traced how everything would fit together. I could continue to try implementing it, but it would help if I could get some guidance on a preferred structure for the implementation.

**Comment by rodrigocfd** [comment]
I never used property sheets before, but as far as I could see, its implementation isn't straightforward... so far, I found [this example](https://learn.microsoft.com/en-us/windows/win32/controls/create-a-property-sheet), is it enough for your needs? Do you have a better one?

I will start implementing [`CreatePropertySheetPage`](https://learn.microsoft.com/en-us/windows/win32/api/prsht/nf-prsht-createpropertysheetpagew) and related stuff.

**Comment by anthonydotmoe** [comment]
I just uploaded a barebones example of making a property sheet page here: [propsheetpage-example](https://github.com/anthonydotmoe/propsheetpage-example).

What I was thinking winsafe could do is give me the same type of abstraction as a normal window or dialog from resource id, except instead of running the dialog directly, it could give me an HPROPSHEETPAGE. The `CreatePropertySheetPage` function really just needs the DlgProc address and the resource ID for the dialog. It could be exactly like in [this example](https://github.com/rodrigocfd/winsafe-examples/blob/master/03_dialog_resources/src/my_modal.rs), except we could expose a `.get_hpropsheetpage` from the `wnd` member.

**Comment by rodrigocfd** [comment]
Yes, I'm think the same: provide a high-level abstraction, reusing the same existing dialog event API. But besides that, the foundation must also exist and be usable on itself, and `PROPSHEETPAGEW` is crazy unsafe.

I'm really careful writing the best possible abstraction around it, and your example will be greatly useful. Thank you very much.

**Comment by rodrigocfd** [comment]
I pushed the preliminar versions (still very unsafe) of [`CreatePropertySheetPage`](https://rodrigocfd.github.io/winsafe/winsafe/prelude/trait.comctl_Hpropsheetpage.html#method.CreatePropertySheetPage) and [`PropertySheet`](https://rodrigocfd.github.io/winsafe/winsafe/fn.PropertySheet.html) functions, along with the structs.

See how much of a lifetime hell this is...

**Comment by anthonydotmoe** [comment]
Wow, thank you for your effort on this request! Yes, I'm finding out the same thing with my MMC project, translating old unsafe C practices to Rust is very difficult. I'll try it out!

**Comment by rodrigocfd** [comment]
> translating old unsafe C practices to Rust is very difficult.

Yes, some APIs are crazy unsafe, and therefore very hard to translate to Rust.

> I'll try it out!

You may try it, but this is just the foundation. It's not finished yet. The next step is build the `gui` abstraction on top of it.
